pool.js

```diff
@@ -296,32 +323,91 @@ Pool.prototype.request_count = function () {
     return count;
 };
 
-Pool.prototype.close = function () {
+Pool.prototype.close = function close() {
     var endpoints = this.endpoints;
     for (var i = 0; i < endpoints.length; i++) {
         endpoints[i].close();
     }
 };
 
-// Dynamic membership
-Pool.prototype.add_endpoint = function (host_port) {
+Pool.prototype.valid_host_port = function valid_host_port(host_port) {
     var ip_port = host_port.split(":");
     var ip = ip_port[0];
     var port = +ip_port[1];
     if (port > 0 && port < 65536) {
-        var endpoint = new GO.PoolEndpoint(this.http, ip, port, this.options);
-        endpoint.on("health", this.endpoint_health_changed.bind(this));
-        endpoint.on("timeout", this.endpoint_timed_out.bind(this));
-        this.endpoints.push(endpoint);
-        this.length++;
-        this.endpoints_by_name[host_port] = endpoint;
+        return ip_port;
     }
+    return null;
 };
 
-Pool.prototype.remove_endpoint = function (host_port) {
-    var ip_port = host_port.split(":");
+// Dynamic membership
+Pool.prototype.add_endpoint = function add_endpoint(host_port) {
+    var ip_port = this.valid_host_port(host_port);
+    if (!ip_port) {
+        return;
+    }
+
+    var hasHostPort = false;
+    for (var i = 0; i < this.all_hostports.length; i++) {
+        if (this.all_hostports[i] === host_port) {
+            hasHostPort = true;
+            break;
+        }
+    }
+
+    if (!hasHostPort) {
+        this.all_hostports.push(host_port);
+    }
+
+    var maxSizeIsRelevant = this.max_pool_size &&
+        this.max_pool_size < this.all_hostports.length;
+
+    // if using maxSize, probabilistically add the new host port to the
+    // active pool endpoints
+    if (maxSizeIsRelevant) {
+        var r = Math.random() * this.all_hostports.length;
+        if (r < this.max_pool_size) {
+            this.add_pool_endpoint(host_port);
```

Comment:
To give you an example of a potential issue, on a fully connected pool you will never add a new hostport because all of the endpoints except the one you just added is unused. This would mean adding new hosts to decrease heavy load on a service would require a restart of the consumers.